### PR TITLE
Implement InputHandler for mouse movement

### PR DESCRIPTION
**Pull request by derekdreery**
This is my state implementation for mouse movement. It adds to PR #181 .

It suits my needs nicely as I can respond to any change in mouse position since the last frame, so I want to see if the behaviour I want is universal.

One particular aspect I invite comments on is defaulting to (0, 0) rather than using an Option. In my model, a movement of (0, 0) is basically a no-op, where as a position of (0, 0) is not a no-op (it's top left window). Does this seem reasonable?

### CODE REVIEW

File: src/ecs/resources/input.rs [hunk 1]
```diff
@@ -13,26 +13,68 @@ enum KeyQueryState {
...
+///
+/// ```ignore
+/// fn handle_events(&mut self,
+///                  events: &[WindowEvent],
+///                  _: &mut World,
+///                  _: &mut AssetManager,
+///                  _: &mut Pipeline)
+///                  -> Trans {
+///     // ...
+///     input_handler.update(events);
```
Comment (Emilgardis) [inline src/ecs/resources/input.rs]: This wouldn't work. Show how to get `input_handler` aswell.

File: src/ecs/resources/input.rs [hunk 2]
```diff
@@ -13,26 +13,68 @@ enum KeyQueryState {
...
+///     // ...
+/// }
+/// ```
 pub struct InputHandler {
     pressed_keys: HashMap<VirtualKeyCode, KeyQueryState>,
+    // use KeyQueryState for mouse button query state as well
+    pressed_mouse_buttons: HashMap<MouseButton, KeyQueryState>,
+    // defined to match glium
+    mouse_position: Option<(i32, i32)>,
+    previous_mouse_position: Option<(i32, i32)>,
```
Comment (Emilgardis) [inline src/ecs/resources/input.rs]: Why does `InputHandler` store the previous mouse position?

File: src/ecs/resources/input.rs [hunk 3]
```diff
@@ -48,30 +90,46 @@ impl InputHandler {
                             // second `Pressed` event.
                         }
                         Entry::Vacant(entry) => {
-                            entry.insert(KeyQueryState::Queried);
+                            entry.insert(KeyQueryState::NotQueried);
```
Comment (Emilgardis) [inline src/ecs/resources/input.rs]: This is wierd, I remember that this was working, but following my logic in the code of key_once it shouldn't work.
Note also that you don't reflect this change in the commit.

**Comment by Emilgardis** [comment]
Also, should #181 be closed inplace of this PR? Seems like they are identical except for `b54a4dc`

**Comment by derekdreery** [comment]
I split the PR up because I thought that the mouse stuff might be more controversial.

File: src/ecs/resources/input.rs (hunk 2 above)
Comment (derekdreery) [inline src/ecs/resources/input.rs]: previous_mouse_position is the mouse position the previous time mouse_change was called. It allows us to compare to the current mouse position and get the amount the mouse has moved. Using this information,  you can work out for example how much to rotate the camera in the ecs etc.

File: src/ecs/resources/input.rs (hunk 1 above)
Comment (derekdreery) [inline src/ecs/resources/input.rs]: Good point!

**Comment by derekdreery** [comment]
I just rebased this on my other PR

**Comment by torkleyy** [comment]
@derekdreery It's not blocked anymore, please rebase onto the current master. I think we can do #225 after this PR is merged.

**Comment by derekdreery** [comment]
This is rebased on current upstream/develop.

**Pull request merged**